ddtrace/llmobs/_integrations/google_genai_utils.py

```diff
@@ -0,0 +1,161 @@
+    "gemini": "google",
+    "imagen": "google",
+    "veo": "google",
+    "jamba": "ai21",
+    "claude": "anthropic",
+    "llama": "meta",
+    "mistral": "mistral",
+    "codestral": "mistral",
+    "deepseek": "deepseek",
+    "olmo": "ai2",
+    "tulu": "ai2",
+    "molmo": "ai2",
+    "specter": "ai2",
+    "cosmoo": "ai2",
+    "qodo": "qodo",
+    "mars": "camb.ai",
+}
+
+
+def extract_provider_and_model_name(kwargs: Dict[str, Any]) -> Tuple[str, str]:
+    model_path = kwargs.get("model", "")
+    model_name = model_path.split("/")[-1]
+    for prefix in KNOWN_MODEL_PREFIX_TO_PROVIDER.keys():
+        if model_name.lower().startswith(prefix):
+            provider_name = KNOWN_MODEL_PREFIX_TO_PROVIDER[prefix]
+            return provider_name, model_name
+    return "custom", model_name if model_name else "custom"
+
+
+def normalize_contents(contents) -> List[Dict[str, Any]]:
+    """
+    contents has a complex union type structure:
+    - contents: Union[ContentListUnion, ContentListUnionDict]
+    - ContentListUnion = Union[list[ContentUnion], ContentUnion]
+    - ContentListUnionDict = Union[list[ContentUnionDict], ContentUnionDict]
+    - ContentUnion = Union[Content, list[PartUnion], PartUnion]
+    - PartUnion = Union[File, Part, str]
+
+    Can also be used for system_instruction which has type ContentUnion
+
+    This function normalizes all these variants into a list of dicts with format{"role": role, "parts": parts}
+    """
+
+    def extract_content(content):
+        role = _get_attr(content, "role", None)
+        parts = _get_attr(content, "parts", None)
+
+        # if parts is missing and content itself is a PartUnion or list[PartUnion]
+        if parts is None:
+            if isinstance(content, list):
+                parts = content
+            else:
+                parts = [content]
+
+        elif not isinstance(parts, list):
+            parts = [parts]
+
+        return {"role": role, "parts": parts}
+
+    if isinstance(contents, list):
+        return [extract_content(c) for c in contents]
+    return [extract_content(contents)]
+
+
+def extract_metrics_google_genai(response) -> Dict[str, Any]:
+    if not response:
+        return {}
+
+    usage_metadata = _get_attr(response, "usage_metadata", {})
+
+    usage = {}
+    input_tokens = _get_attr(usage_metadata, "prompt_token_count", None)
+    output_tokens = _get_attr(usage_metadata, "candidates_token_count", None)
+    cached_tokens = _get_attr(usage_metadata, "cached_content_token_count", None)
+    total_tokens = _get_attr(usage_metadata, "total_token_count", None) or input_tokens + output_tokens
+
+    if input_tokens is not None:
+        usage[INPUT_TOKENS_METRIC_KEY] = input_tokens
+    if output_tokens is not None:
+        usage[OUTPUT_TOKENS_METRIC_KEY] = output_tokens
+    if cached_tokens is not None:
+        usage["cached_tokens"] = cached_tokens  # TODO(max): add constant for cached tokens when it is introduced
```

Comment:
NVM looks like Evan's PR is blocked. Let's add this constant in this PR: https://github.com/DataDog/dd-trace-py/pull/13755/files#diff-39d497917e6f807b9e71a057dd8da8a852debec65c78c52b412cdc492e7bdf56R41